integration-tests/src/test/java/com/datastax/oss/driver/mapper/ImmutableIT.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.mapper;
+
+import static com.datastax.oss.driver.api.mapper.entity.naming.GetterStyle.SHORT;
+
+import com.datastax.oss.driver.api.mapper.annotations.Entity;
+import com.datastax.oss.driver.api.mapper.annotations.IntrospectionStrategy;
+import com.datastax.oss.driver.api.mapper.annotations.PartitionKey;
+import java.util.UUID;
+
+public class ImmutableIT extends InventoryITBase {
+
+  @Entity
+  @IntrospectionStrategy(getterStyle = SHORT, mutable = false)
+  public static class ImmutableProduct {
```

Comment:
No tests yet, but the generated entity helper looks correct.

Reply:
I may not be qualified to do suggestions here, but I am using the driver from kotlin and in this context I was excitedly following this Issue and pull request.

Couldn't the mutabillity be infered?
 - Search for getters => extract member names
 - Search for setters => extract member names
 - mutable = getters.memberNames == setters.memberNames 

This would make all kotlin and java immutables compatible out of the box, without providing a specification via annotation. 

--
In this context java dropped the classical getter/setter naming convention in their [record types](https://openjdk.java.net/jeps/359) aswell. To say so with the introduction of java 14, anyone using records has to use this annotation too, but I have no idea how one could tackle this issue.

Reply:
Hi,

Community feedback is definitely welcome 😄 

I thought about auto-detection but it seems hard to come up with a reliable set of rules. Consider the following edge cases:

```java
// Uses Javabeans getter style, but there is also a random method that masquerades as a short-style getter
// If I encounter `int a()` first, I could legitimately assume that I'm dealing with a short-style immutable entity.
@Entity class Product {
  private int id;
  public int a() { return 42; };
  public int getId() { return id; }
  public void setId(int id) { this.id = id; }
  }
}

// Short-style getter, but the property name happens to look like a Javabeans getter.
// When I encounter `getMethodName()`, I could legitimately assume that I'm dealing with a Javabeans-style
// immutable entity that has a property called `methodName`.
@Entity class Field {
  private String getMethodName;
  public String getMethodName() { return getMethodName; }
  public void setGetMethodName(String getMethodName) { this.getMethodName = getMethodName; }
}
```
These are definitely convoluted cases, but since we're dealing with arbitrary input we must deal with them. Knowing the strategy beforehand makes this trivial, we can take decisions by looking at each method in isolation. It does require explicit annotation by the user, but I think that's a small price to pay since you can put the `@PropertyStrategy` annotation on a parent interface and inherit it on all your entities. We can also make educated guesses for specific languages: the PR already does so with `scala.Product` for case classes, I will check if what can be done for Kotlin data classes.

Thanks for bringing up Java 14 records too, I hadn't thought about that.

Reply:
Good news, there's a `kotlin.Metadata` annotation that gets added to every Kotlin-generated class since 1.3, and it's visible from the annotation processor. We should probably default to `mutable=false`, because that's how Kotlin users are likely using data classes.

I'll revisit the code to introduce a more generic language detection mechanism.

Reply:
Kotlin data classes use the java bean getter style, so a mutable/immutable autodetection would be enough and totaly posible if you specified the getter/setter style, since you then know the column names. Detecting the getter style could indeed be problematic, you must either juggle around with the different types of getter/setters/constructors and find a fitting one or find a different ground trouth for the column names, so one can check if either the bean style or the short getters or setters are "column complete".

Reply:
for java 14 records it will be the "java/lang/Record" supertype, so language detection would work there too

Reply:
I also tested the mapper with kotlin data classes in one of our projects (thanks to jitpack) and they seem to work fine.

Reply:
Yes, Kotlin works particularly well since it's tightly integrated with the annotation processing mechanism. I don't know if you were aware of it, but we're already covering it in the [manual](https://docs.datastax.com/en/developer/java-driver/4.8/manual/mapper/config/#kotlin). The only thing that will change with this PR is that immutability will become the default, I think it makes sense because it's what people will expect with data classes.

I get your point about inferring mutability, but I think there would still be ambiguous situations, e.g. if one property has a setter, but another doesn't. One nice aspect of the current model is that we can immediately take a decision as we process each individual method. If we introduce rules like "if at least one property has a setter, then the entity is mutable and we should ignore properties that don't have setters", then we'd have to do stuff in two passes, I'm not sure if it's worth adding that complexity.

> for java 14 records it will be the "java/lang/Record" supertype, so language detection would work there too

👍 
Done. There's just a small hack when we validate the annotation's target, because they introduced a new `ElementKind.RECORD`. But apart from that it's transparent for the rest of the processor.

Reply:
> I get your point about inferring mutability, but I think there would still be ambiguous situations, e.g. if one property has a setter, but another doesn't. One nice aspect of the current model is that we can immediately take a decision as we process each individual method. If we introduce rules like "if at least one property has a setter, then the entity is mutable and we should ignore properties that don't have setters", then we'd have to do stuff in two passes, I'm not sure if it's worth adding that complexity.

Agree.